From: &lt;gshchepaDate: August 23 2007 10:54pm
Subject: bk commit into 5.1 tree (gshchepa:1.2569) BUG#30287

Below is the list of changes that have just been committed into a local
5.1 repository of uchum. When uchum does a push these changes will
be propagated to the main repository and, within 24 hours after the
push, to the public repository.
For information on how to access the public repository
see http://dev.mysql.com/doc/mysql/en/installing-source-tree.html

ChangeSet@stripped, 2007-08-24 01:54:18+05:00, gshchepa@stripped +3 -0
  Fixed bug #30287.
  Recommit to 5.1.22.
  The server created temporary tables for filesort in the working directory
  instead of the specified tmpdir directory.

  sql/item.cc@stripped, 2007-08-24 01:52:11+05:00, gshchepa@stripped +2 -0
    Fixed bug #30287.
    Recommit to 5.1.22.
    The Item_field::set_field method has been modified to reset the any_privileges
    flag to false in case of system temporary table. This modification prevents the
    server from unnecessary checking of user privileges to access system temporary
    tables.

  sql/sql_select.cc@stripped, 2007-08-24 01:52:15+05:00, gshchepa@stripped +8 -10
    Fixed bug #30287.
    Recommit to 5.1.22.
    Bugfix for #29015 has been removed: TABLE_SHARE::table_name of system
    temporary tables contains full path to table file basename again.

  sql/sql_view.cc@stripped, 2007-08-24 01:52:16+05:00, gshchepa@stripped +6 -0
    Fixed bug #30287.
    Recommit to 5.1.22.
    Commentary has been added.

diff -Nrup a/sql/item.cc b/sql/item.cc
--- a/sql/item.cc	2007-08-14 00:39:23 +05:00
+++ b/sql/item.cc	2007-08-24 01:52:11 +05:00
@@ -1801,6 +1801,8 @@ void Item_field::set_field(Field *field_
   unsigned_flag=test(field_par->flags & UNSIGNED_FLAG);
   collation.set(field_par->charset(), field_par->derivation());
   fixed= 1;
+  if (field->table->s->tmp_table == SYSTEM_TMP_TABLE)
+    any_privileges= 0;
 }
 
 
diff -Nrup a/sql/sql_select.cc b/sql/sql_select.cc
--- a/sql/sql_select.cc	2007-08-14 00:39:25 +05:00
+++ b/sql/sql_select.cc	2007-08-24 01:52:15 +05:00
@@ -9417,7 +9417,7 @@ create_tmp_table(THD *thd,TMP_TABLE_PARA
   bool  using_unique_constraint= 0;
   bool  use_packed_rows= 0;
   bool  not_all_columns= !(select_options & TMP_TABLE_ALL_COLUMNS);
-  char	*tmpname, *tmppath, path[FN_REFLEN], table_name[NAME_LEN+1];
+  char  *tmpname,path[FN_REFLEN];
   uchar	*pos, *group_buff, *bitmaps;
   uchar *null_flags;
   Field **reg_field, **from_field, **default_field;
@@ -9441,12 +9441,12 @@ create_tmp_table(THD *thd,TMP_TABLE_PARA
     temp_pool_slot = bitmap_lock_set_next(&temp_pool);
 
   if (temp_pool_slot != MY_BIT_NONE) // we got a slot
-    sprintf(table_name, "%s_%lx_%i", tmp_file_prefix,
+    sprintf(path, "%s_%lx_%i", tmp_file_prefix,
             current_pid, temp_pool_slot);
   else
   {
     /* if we run out of slots or we are not using tempool */
-    sprintf(table_name, "%s%lx_%lx_%x", tmp_file_prefix,current_pid,
+    sprintf(path,"%s%lx_%lx_%x", tmp_file_prefix,current_pid,
             thd->thread_id, thd->tmp_table++);
   }
 
@@ -9454,8 +9454,8 @@ create_tmp_table(THD *thd,TMP_TABLE_PARA
     No need to change table name to lower case as we are only creating
     MyISAM or HEAP tables here
   */
-  fn_format(path, table_name, mysql_tmpdir, "",
-            MY_REPLACE_EXT|MY_UNPACK_FILENAME);
+  fn_format(path, path, mysql_tmpdir, "", MY_REPLACE_EXT|MY_UNPACK_FILENAME);
+
 
   if (group)
   {
@@ -9501,8 +9501,7 @@ create_tmp_table(THD *thd,TMP_TABLE_PARA
                         sizeof(*key_part_info)*(param->group_parts+1),
                         &param->start_recinfo,
                         sizeof(*param->recinfo)*(field_count*2+4),
-                        &tmppath, (uint) strlen(path)+1,
-                        &tmpname, (uint) strlen(table_name)+1,
+                        &tmpname, (uint) strlen(path)+1,
                         &group_buff, (group && ! using_unique_constraint ?
                                       param->group_length : 0),
                         &bitmaps, bitmap_buffer_size(field_count)*2,
@@ -9521,8 +9520,7 @@ create_tmp_table(THD *thd,TMP_TABLE_PARA
     DBUG_RETURN(NULL);				/* purecov: inspected */
   }
   param->items_to_copy= copy_func;
-  strmov(tmppath, path);
-  strmov(tmpname, table_name);
+  strmov(tmpname,path);
   /* make table according to fields */
 
   bzero((char*) table,sizeof(*table));
@@ -9547,7 +9545,7 @@ create_tmp_table(THD *thd,TMP_TABLE_PARA
   table->keys_in_use_for_query.init();
 
   table->s= share;
-  init_tmp_table_share(share, "", 0, tmpname, tmppath);
+  init_tmp_table_share(share, "", 0, tmpname, tmpname);
   share->blob_field= blob_field;
   share->blob_ptr_size= mi_portable_sizeof_char_ptr;
   share->db_low_byte_first=1;                // True for HEAP and MyISAM
diff -Nrup a/sql/sql_view.cc b/sql/sql_view.cc
--- a/sql/sql_view.cc	2007-07-17 00:02:46 +05:00
+++ b/sql/sql_view.cc	2007-08-24 01:52:16 +05:00
@@ -397,7 +397,13 @@ bool mysql_create_view(THD *thd, TABLE_L
     {
       Item_field *field;
       if ((field= item->filed_for_view_update()))
+      {
+        /*
+         any_privileges may be reset later by the Item_field::set_field
+         method in case of a system temporary table.
+        */
         field->any_privileges= 1;
+      }
     }
   }
 #endif
